### PR TITLE
Use ES modules in server sources

### PR DESCRIPTION
**Pull request by jsnajdr**
Many modules in the Calypso server (`client/server`) use a questionable mix of ESM and CJS syntax:
```js
import fs from 'fs';

module.exports = function () {
  ...
}
```
This works only because the Babel config for server build [specifies](https://github.com/Automattic/wp-calypso/blob/b4e8fbb6cd9852d3017164593f2c170b9a8abc15/babel.config.js#L4) `modules: 'commonjs'`, transpiles ESM into CJS and webpack never sees a ES module source.

If webpack saw the mixed ESM/CJS module, assigning to `module.exports` would fail at runtime:
```
Cannot assign to read-only property 'exports' of object ...
```

This PR converts most of the hybrid server modules to pure ESM and removes the now redundant Babel config flag. By default (from `calypso-build`), the value will be `modules: false`, meaning "no transpilation of `import`/`export` syntax). Webpack will handle the modules when bundling.

Some modules need to remain CJS, because they are also used by the build system, i.e., Babel or webpack. Typically, it's the `config` module.

When importing an ESM module with default export with `require()`, one needs to be careful to use the `.default` field to access the default export.

See also https://github.com/Automattic/wp-desktop/pull/884#issuecomment-633522275 and https://github.com/Automattic/wp-desktop/pull/887

**How to test:**
1. Test that devdocs continue to work, including selectors search
2. Test that OAuth continues to work (in `wp-desktop`)
3. Test the "sign in with Apple" feature, as we modified the redirect API router for that (@Tug can you help with this testing? Either directly or by providing advice on how to best test this part)
4. Test the user bootstrapping (will need to happen on staging right before deploy)

### CODE REVIEW

**Comment by sgomes** [comment]
Thanks, @jsnajdr! Having some consistency is nice 🙂 

The changes generally look good, but I'm not so sure about having inline `require` statements for conditionals in otherwise ESM files:

```js
if ( config.isEnabled( 'oauth' ) && ! config.isEnabled( 'jetpack-cloud' ) ) {
  require( './oauth' ).default( app );
}
```

ESM doesn't really offer a good alternative to this, since dynamic imports are async. Should those files stay as CommonJS, so that everything is a `require`?

**Comment by jsnajdr** [comment]
> I'm not so sure about having inline `require` statements for conditionals

As you write, ESM doesn't offer any synchronous alternative for that.

The `require()` statements are harmless, so we can just leave them there.

Or we can simply `import` the modules unconditionally, and only call the imported functions conditionally.

Doing the import conditionally doesn't really add that much value. It speeds up server startup a little bit, for certain configurations, by avoiding parsing and execution of the modules.

We need to check that the modules don't do any static initialization, or in other words, execute stuff only when the exported functions are called.

What do you think? I'd lean towards always doing the `import`s and moving them to the top.

**Comment by sgomes** [comment]
> What do you think? I'd lean towards always doing the `import`s and moving them to the top.

That sounds quite reasonable, yes. It's what would most likely happen if this were client code 👍 

I also agree with the point about static initialization. We need to make sure that there aren't any undesirable side effects to loading the modules unconditionally.

**Comment by Tug** [comment]
Tested Sign in with Apple on this branch, all good 👍

**Comment by nsakaimbo** [comment]
Change seems reasonable, but it appears there's a runtime error with the desktop app? ([link: workflow](https://app.circleci.com/pipelines/github/Automattic/wp-desktop/14448/workflows/157688a0-3f55-45c4-b1da-1264131f4c95/jobs/62431/steps))

[Video artifact](https://app.circleci.com/pipelines/github/Automattic/wp-desktop/14448/workflows/157688a0-3f55-45c4-b1da-1264131f4c95/jobs/62431/artifacts) of the test run shows the error: `t is not a function...`

Screen Shot 2020-05-26 at 11 55 29 AM

**Review by nsakaimbo — DISMISSED** [review]
Changes make sense and consistent use of ES modules seems like a necessary change. However looks like there's a runtime issue with the desktop that needs to be resolved (see comment above).

**Comment by jsnajdr** [comment]
> but it appears there's a runtime error with the desktop app?

Thanks for spotting this @nsakaimbo! The conversion of `server/boot` breaks the `require( 'server/boot' )` statement in `wp-desktop`. Fixing in https://github.com/Automattic/wp-desktop/pull/890 which should be merged to `wp-desktop` before merging this one to `wp-calypso`.

**Comment by jsnajdr** [comment]
@sgomes I converted some conditional `require()` calls to static `import`s in https://github.com/Automattic/wp-calypso/pull/42643/commits/12929edb046469a50ec217f37e79b718f6570c5e

Some `require`s need to stay, for example, the `server/devdocs` one. The `server/devdocs` module imports the `search-index` generated file, which is not present in `wp-desktop` build and would cause error at boot. I'd like to eventually resolve this in another PR.

**Review by sgomes — APPROVED** [review]
Thanks, @jsnajdr! 

The changes look good to me, assuming the `wp-desktop` issue gets solved separately 👍

**Comment by sgomes** [comment]
Added the `do not merge` label while we're waiting for the `wp-desktop` fix.

**Comment by jsnajdr** [comment]
All checks succeeded, going to merge now 🎉

**Pull request merged**
